### PR TITLE
consolidate argument specs and doc fragments

### PR DESCRIPTION
**Pull request by mikemorency**
##### SUMMARY
Combining https://github.com/ansible-collections/vmware.vmware/pull/86 and https://github.com/ansible-collections/vmware.vmware/pull/87

This removes instances of repeated option documentation and redundant spec args.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
doc fragments
argument specs

##### ADDITIONAL INFORMATION
While I was working on this, I found the kubernetes collection has a pretty nice pattern for breaking out the doc fragments. I tried to copy theirs, https://github.com/ansible-collections/kubernetes.core/tree/main/plugins/doc_fragments
Similar pattern for the arg specs too https://github.com/ansible-collections/kubernetes.core/blob/main/plugins/module_utils/args_common.py

### CODE REVIEW

File: plugins/doc_fragments/additional_rest_options.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    # that use the REST API SDK. You must include the base fragment in addition to this
+    #
+    # This vmware.vmware.additional_rest_options fragment will cover any options returned by rest_compatible_argument_spec()
+    # that are not included in vmware.vmware.base
+    DOCUMENTATION = r'''
+options:
+  proxy_protocol:
+    description:
+      - The proxy connection protocol to use.
+      - This is option is used if the correct proxy protocol cannot be automatically determined.
```
Comment (anna-savina) [inline plugins/doc_fragments/additional_rest_options.py:24]:
```suggestion
      - This option is used if the correct proxy protocol cannot be automatically determined.
```

File: plugins/doc_fragments/base_options.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class ModuleDocFragment(object):
+    # This document fragment serves as a base for all vmware modules. If you are using the REST API SDK in your module,
+    # you should also include the vmware.vmware.additional_rest_options fragment.
+    #
+    # This vmware.vmware.base_options fragment covers the arg spec provided by the base_argument_spec() function
+    DOCUMENTATION = r'''
+notes:
+  - All modules require API write access and hence is not supported on a free ESXi license.
```
Comment (anna-savina) [inline plugins/doc_fragments/base_options.py:20]:
```suggestion
  - All modules require API write access and hence are not supported on a free ESXi license.
```

File: plugins/doc_fragments/base_options.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    default: true
+  port:
+    description:
+      - The port number of the vSphere vCenter server.
+      - If the value is not specified in the task, the value of environment variable E(VMWARE_PORT) will be used instead.
+    type: int
+    default: 443
+  proxy_host:
+    description:
+      - Address of a proxy that will receive all HTTPS requests and relay them.
```
Comment (anna-savina) [inline plugins/doc_fragments/base_options.py:58]:
```suggestion
      - The address of a proxy that will receive all HTTPS requests and relay them.
```

File: plugins/doc_fragments/base_options.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      - The format is a hostname or a IP.
+      - If the value is not specified in the task, the value of environment variable E(VMWARE_PROXY_HOST) will be used instead.
+    type: str
+    required: false
+  proxy_port:
+    description:
+    - Port of the HTTP proxy that will receive all HTTPS requests and relay them.
```
Comment (anna-savina) [inline plugins/doc_fragments/base_options.py:65]:
```suggestion
    - The port of the HTTP proxy that will receive all HTTPS requests and relay them.
```

File: plugins/module_utils/_vmware_argument_spec.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+from ansible.module_utils.basic import env_fallback
+
+
+def rest_compatible_argument_spec():
+    """
+    This returns a dictionary that can be used as the baseline for all REST module specs.
+    If your module uses the REST API, you should use this instead of the base_argument_spec.
+    If your module uses both this and the pyvmomi SDK, you should still use this spec.
```
Comment (anna-savina) [inline plugins/module_utils/_vmware_argument_spec.py:8]:
```suggestion
    If your module uses both this spec and the pyvmomi SDK, you should still use this spec.
```

File: plugins/module_utils/_vmware_argument_spec.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                choices=['https', 'http'],
+                aliases=['protocol']
+            ),
+        )
+    }
+
+
+def base_argument_spec():
+    """
+    This returns a dictionary that can be used as the baseline for all vmware module specs. Any arguments
```
Comment (anna-savina) [inline plugins/module_utils/_vmware_argument_spec.py:25]:
```suggestion
    This returns a dictionary that can be used as the baseline for all VMware module specs. Any arguments
```

**Review by anna-savina — APPROVED** [review]
LGTM

**Review by mariolenz — COMMENTED** [review]
@mikemorency I'd like to test your changes with `community.vmware`. That is, what would happen if I try to (re-) use them there.

Please give me a few days for this.

**Comment by mikemorency** [comment]
Sure thing @mariolenz, I appreciate the additional testing. If possible, I'd like to merge by the 18th so we can do some additional QA before the targeted release on the 23rd

File: plugins/module_utils/_vmware.py [hunk 7]
```diff
@@ -32,54 +32,14 @@
...
-from ansible.module_utils.basic import env_fallback, missing_required_lib
+from ansible.module_utils.basic import missing_required_lib
 
 
 class ApiAccessError(Exception):
     def __init__(self, *args, **kwargs):
         super(ApiAccessError, self).__init__(*args, **kwargs)
 
 
-def vmware_argument_spec():
```
Comment (mariolenz) [inline plugins/module_utils/_vmware.py:43]: @mikemorency Why didn't you also remove `vmware_client_argument_spec()` from `plugins/module_utils/_vmware_rest_client.py`?

**Review by mariolenz — APPROVED** [review]
@mikemorency Apart from my last comment above LGTM

_ref_ ansible-collections/community.vmware#2273

File: plugins/module_utils/_vmware.py (hunk 7 above)
Comment (mikemorency) [reply plugins/module_utils/_vmware.py:43]: Good catch/suggestion!

I tried to not touch the rest client initially, and just stick to pyvmomi. Once i split everything up it made less sense to leave the rest client stuff alone for the docs/arg specs. I missed this, ill remove it

**Pull request merged**